pkg/services/notifications/testing.go

```diff
@@ -0,0 +1,5 @@
+package notifications
+
+func (ns *NotificationService) MailQueuePop() *Message {
```

Comment:
I was slightly unsure what the best approach was for exposing functionality to tests but not otherwise. I certainly didn't want to export the mail queue channel alone. If there is a better way to expose functionality to tests, I'm all for it.

(Perhaps this is an indicator that the fake route mentioned in the PR description is better?)

Reply:
I really don't think adding a public method to the `NotificationService` just for testing purposes is a good idea.

Reply:
Agreed - I'm open to suggestions on how to best to expose something to tests in Go without exposing it publicly.

Other languages I've worked with (C#, for instance) has a mechanism for exposing non-public APIs to tests for exactly this purpose. I felt that the method was cleaner than exposing the raw channel directly, since it at least limits the things a caller can do with it.

The main issue is that `SendEmailCommandSync` reaches out and tries to dial a real external SMTP server. In a test, we need to intercept this, and the current notification service provides no hooks to do so aside from this channel. The direct use of the channel is what [another similar integration test](https://github.com/grafana/grafana/blob/main/pkg/services/notifications/send_email_integration_test.go) already does, but since this test is in the notifications package, it can access the private channel directly. Our test isn't in this package, unfortunately.

The other approach (which might be better long-term, honestly) would be to refactor the core email service to use an interface in front of [the SMTP dial](https://github.com/grafana/grafana/blob/c3eb1ffe850be4f3ee4093987160fbb29e196e2f/pkg/services/notifications/mailer.go#L63). Then, we can add a fake which intercepts the message. I originally didn't go this route because I felt it was too invasive of a change, but if we think otherwise I'm happy to do it. It'd allow us to continue using the `Sync` command.

What are your thoughts on this @santihernandezc ? Do you know of some Go magic that'd let me access this in a test but not expose it publicly? Otherwise, would the more invasive change described above ease your concerns?

Reply:
Got it. Makes sense, Alex.
There's not any Go magic or hacks to expose a private field/function that doesn't fall under the "anti-pattern" category, at least as far as I know! When you have an external test you can only reach what's publicly exposed by the package being used.
IMHO we could get away with using this `MailQueuePop()` function for now, but we should open a related issue to refactor the email service and continue using the `Sync` command as soon as this gets merged.

Reply:
Trying out the alternative proposed above, after which we can avoid the temporary implementation and stick with synchronicity. See #43875

Reply:
Done, now that the proper abstractions were added in another PR. We no longer depend on the async version and we are back to using the sync, but the tests still work.